adagrams/game.py

```diff
@@ -1,13 +1,138 @@
+        letter = letter_list[index]
+        count = letter_count_dict[letter] 
+        hand.append(letter)
+        letter_count_dict[letter] = count - 1
+        # new_count = count  - 1
+        if count == 1:
+            letter_list.pop(index)
+
+    return hand
 
 def uses_available_letters(word, letter_bank):
-    pass
+    """
+    Check if an input word from a user only used characters that are within the hand list
+    Parameters:
+    - word as a string
+    - letter_bank as a list of drawn letters in a hand
+    Return: boolean
+    """
+    # create letter_freq_dict for letters in letter_bank
+    letter_freq_dict = {}
+    for letter in letter_bank:
+        if letter in letter_freq_dict:
+            letter_freq_dict[letter] += 1
+        else:
+            letter_freq_dict[letter] = 1
+    # iterate through each character in word string and check if char exists and compare counts; update counts when finding a matching
+    for char in word.upper():
+        if not char in letter_freq_dict or letter_freq_dict[char] == 0:
+            return False
+        else:
+            letter_freq_dict[char] -= 1
+    return True
 
 def score_word(word):
-    pass
+    """
+    Score the word from user input
+    Parameters: word as a string of chars
+    output: int (number of points)    
+    """
+    # create LETTER_VALUES as a global variable outside (see top)
+    # iterate through word string, find points for each char and add them up, check length for extra points
+    score = 0
+    for char in word.upper():
+        score += LETTER_VALUES[char]
+    if len(word) in [7, 8, 9, 10]:
+        score += 8
+    return score
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    """
+    Find word with highest score
+    Parameter: a list of strings
+    Output: a tuple with a string and the score    
+    """
+    # calculate score for each word and store in word_scores dict, also update highest score
+    word_scores = {}
+    highest_score = 0
+    for word in word_list:
+        score = score_word(word)
+        word_scores[word] = score
+        if score > highest_score:
+            highest_score = score
+    # loop through the word_list, find all words with highest score
+    words_with_highest_score = []
+    for word in word_list:
+        if word_scores[word] == highest_score:
+            words_with_highest_score.append(word)
+    
+    winning_word = words_with_highest_score[0]
+    # return if no ties
+    if len(words_with_highest_score) == 1:
+        return (winning_word, highest_score)
+    # find the shortest word, also check if length of 10 exists    
+    smallest_length = len(winning_word)
+    for word in words_with_highest_score:
+        if len(word) == 10:
+            return (word, highest_score)
+        if len(word) < smallest_length:
+            winning_word = word
+            smallest_length = len(word)
+    return (winning_word, highest_score)
```

Comment:
These tie breakers here are handled beautifully. I like how up to this point, you have filtered out everything that can be easily ruled out as the winning word. For this final loop, you only look at the words that have the highest score!